File: mobile/novalang_flutter/lib/core/utils/native_content.dart
String normalizeNativeLocale(String languageCode) =>
    languageCode.trim().toLowerCase().replaceAll('_', '-').split('-').first;

String missingNativeContentSentinel(String path, String languageCode) =>
    '⟦missing-content:$path:${normalizeNativeLocale(languageCode)}⟧';

String strictNativeText(
  Map<String, String> values,
  String languageCode, {
  required String path,
  String legacy = '',
}) {
  final locale = normalizeNativeLocale(languageCode);
  if (values.isEmpty) {
    return legacy.trim().isEmpty
        ? missingNativeContentSentinel(path, locale)
        : legacy;
  }
  final value = values[locale]?.trim();
  return value == null || value.isEmpty
      ? missingNativeContentSentinel(path, locale)
      : value;
}

List<String> strictNativeTextList(
  Map<String, List<String>> values,
  String languageCode, {
  required String path,
  List<String> legacy = const [],
}) {
  final locale = normalizeNativeLocale(languageCode);
  if (values.isEmpty) {
    return legacy.isEmpty
        ? [missingNativeContentSentinel(path, locale)]
        : legacy;
  }
  final value = values[locale];
  if (value == null ||
      value.isEmpty ||
      value.any((item) => item.trim().isEmpty)) {
    return [missingNativeContentSentinel(path, locale)];
  }
  return value;
}

/// Resolves field-level `*ByNative` maps without crossing language boundaries.
///
/// Shared validators guarantee that learner-support fields contain vi/en/ja.
/// The sentinel keeps a malformed development asset visible without silently
/// substituting English or Vietnamese in production UI.
Map<String, dynamic> resolveNativeContentMap(
  Map<String, dynamic> source,
  String languageCode, {
  String path = 'content',
}) {
  final locale = normalizeNativeLocale(languageCode);
  final resolved = <String, dynamic>{};

  for (final entry in source.entries) {
    if (entry.key.endsWith('ByNative')) continue;
    final localized = source['${entry.key}ByNative'];
    resolved[entry.key] = localized is Map
        ? _resolveLocalizedValue(localized, locale, '$path.${entry.key}')
        : _resolveNested(entry.value, locale, '$path.${entry.key}');
  }

  for (final entry in source.entries.where(
    (entry) => entry.key.endsWith('ByNative'),
  )) {
    final baseKey = entry.key.substring(
      0,
      entry.key.length - 'ByNative'.length,
    );
    if (resolved.containsKey(baseKey)) continue;
    final localized = entry.value;
    if (localized is Map) {
      resolved[baseKey] = _resolveLocalizedValue(
        localized,
        locale,
        '$path.$baseKey',
      );
    }
  }

  return resolved;
}

dynamic _resolveLocalizedValue(
  Map<dynamic, dynamic> values,
  String locale,
  String path,
) {
  final value = values[locale];
  if (value == null || (value is String && value.trim().isEmpty)) {
    return missingNativeContentSentinel(path, locale);
  }
  return _resolveNested(value, locale, path);
}

dynamic _resolveNested(dynamic value, String locale, String path) {
  if (value is Map) {
    return resolveNativeContentMap(
      value.map((key, item) => MapEntry(key.toString(), item)),
      locale,
      path: path,
    );
  }
  if (value is List) {
    return List<dynamic>.generate(
      value.length,
      (index) => _resolveNested(value[index], locale, '$path[$index]'),
      growable: false,
    );
  }
  return value;
}

File: frontend/src/utils/nativeContent.ts
/** Port of Flutter `resolveNativeContentMap` — collapses `*ByNative` maps to the active locale. */

export const MISSING_CONTENT_PATTERN = /⟦missing(?:-content)?:[^⟧]*⟧/g;

export function normalizeNativeLocale(languageCode: string): string {
  return languageCode.trim().toLowerCase().replace(/_/g, "-").split("-")[0] ?? "en";
}

export function missingNativeContentSentinel(path: string, languageCode: string): string {
  return `⟦missing-content:${path}:${normalizeNativeLocale(languageCode)}⟧`;
}

export function isMissingContentSentinel(value: unknown): boolean {
  return typeof value === "string" && /⟦missing(?:-content)?:/.test(value);
}

/** Strip missing-content sentinels for end-user display (show empty instead). */
export function displayNativeText(value: unknown): string {
  if (typeof value !== "string") return "";
  const trimmed = value.trim();
  if (!trimmed || isMissingContentSentinel(trimmed)) return "";
  return trimmed.replace(MISSING_CONTENT_PATTERN, "").trim();
}

export function strictNativeText(
  values: Record<string, string> | null | undefined,
  languageCode: string,
  path: string,
  legacy = "",
): string {
  const locale = normalizeNativeLocale(languageCode);
  if (!values || Object.keys(values).length === 0) {
    return legacy.trim() ? legacy : missingNativeContentSentinel(path, locale);
  }
  const value = values[locale]?.trim();
  return value ? value : missingNativeContentSentinel(path, locale);
}

function resolveLocalizedValue(
  values: Record<string, unknown>,
  locale: string,
  path: string,
): unknown {
  const value = values[locale];
  if (value == null || (typeof value === "string" && value.trim() === "")) {
    return missingNativeContentSentinel(path, locale);
  }
  return resolveNested(value, locale, path);
}

function resolveNested(value: unknown, locale: string, path: string): unknown {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return resolveNativeContentMap(value as Record<string, unknown>, locale, path);
  }
  if (Array.isArray(value)) {
    return value.map((item, index) => resolveNested(item, locale, `${path}[${index}]`));
  }
  return value;
}

/**
 * Resolves field-level `*ByNative` maps without crossing language boundaries.
 * Same semantics as Flutter `resolveNativeContentMap`.
 */
export function resolveNativeContentMap(
  source: Record<string, unknown>,
  languageCode: string,
  path = "content",
): Record<string, unknown> {
  const locale = normalizeNativeLocale(languageCode);
  const resolved: Record<string, unknown> = {};

  for (const [key, entryValue] of Object.entries(source)) {
    if (key.endsWith("ByNative")) continue;
    const localized = source[`${key}ByNative`];
    resolved[key] =
      localized && typeof localized === "object" && !Array.isArray(localized)
        ? resolveLocalizedValue(localized as Record<string, unknown>, locale, `${path}.${key}`)
        : resolveNested(entryValue, locale, `${path}.${key}`);
  }

  for (const [key, entryValue] of Object.entries(source)) {
    if (!key.endsWith("ByNative")) continue;
    const baseKey = key.slice(0, -"ByNative".length);
    if (Object.prototype.hasOwnProperty.call(resolved, baseKey)) continue;
    if (entryValue && typeof entryValue === "object" && !Array.isArray(entryValue)) {
      resolved[baseKey] = resolveLocalizedValue(
        entryValue as Record<string, unknown>,
        locale,
        `${path}.${baseKey}`,
      );
    }
  }

  return resolved;
}

export function asStringList(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => displayNativeText(item))
    .filter((item) => item.length > 0);
}

export function asRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

export function asRecordList(value: unknown): Record<string, unknown>[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => asRecord(item))
    .filter((item): item is Record<string, unknown> => item != null);
}
